apps/openmw/mwrender/sky.cpp

```diff
@@ -336,42 +336,85 @@ class CelestialBody
         mTransform->addChild(mGeode);
 
         parentNode->addChild(mTransform);
+
+        mGeode->addUpdateCallback(this);
     }
 
-    void setDirection(const osg::Vec3f& direction)
+    virtual ~CelestialBody()
     {
-        osg::Vec3f normalizedDirection = direction / direction.length();
-        mTransform->setPosition(normalizedDirection*1000.f);
-
-        osg::Quat quat;
-        quat.makeRotate(osg::Vec3f(0,0,1), normalizedDirection);
-        mTransform->setAttitude(quat);
+        mGeode->removeUpdateCallback(this);
     }
 
+    virtual void adjustTransparency(const float ratio) = 0;
+
     void setVisible(bool visible)
     {
         mTransform->setNodeMask(visible ? ~0 : 0);
     }
 
 protected:
+    static const float mDistance;
     osg::ref_ptr<osg::PositionAttitudeTransform> mTransform;
     osg::ref_ptr<osg::Geode> mGeode;
-    Resource::SceneManager* mSceneManager;
-
 };
 
+const float CelestialBody::mDistance = 1000.0f;
+
 class Sun : public CelestialBody
 {
 public:
-    Sun(osg::Group* parentNode, Resource::SceneManager* sceneManager)
-        : CelestialBody(parentNode, sceneManager, 1.f, 1)
+    Sun(osg::Group* parentNode, Resource::TextureManager& textureManager)
+        : CelestialBody(parentNode, 1.0f, 1)
+        , mColor(1.0f, 1.0f, 1.0f, 1.0f)
+        , mDummyTexture(textureManager.getWarningTexture())
     {
-        osg::ref_ptr<osg::Texture2D> tex = mSceneManager->getTextureManager()->getTexture2D("textures/tx_sun_05.dds",
-                                                                                               osg::Texture::CLAMP, osg::Texture::CLAMP);
+        osg::ref_ptr<osg::Texture2D> tex = textureManager.getTexture2D("textures/tx_sun_05.dds",
+                                                                       osg::Texture::CLAMP,
+                                                                       osg::Texture::CLAMP);
 
         mTransform->getOrCreateStateSet()->setTextureAttributeAndModes(0, tex, osg::StateAttribute::ON);
         mTransform->getOrCreateStateSet()->setAttributeAndModes(createUnlitMaterial(), osg::StateAttribute::ON);
     }
+
+    virtual void setDefaults(osg::StateSet* stateset)
+    {
+        osg::ref_ptr<osg::TexEnvCombine> texEnv(new osg::TexEnvCombine);
+        texEnv->setCombine_Alpha(osg::TexEnvCombine::MODULATE);
+        texEnv->setSource0_Alpha(osg::TexEnvCombine::PREVIOUS);
+        texEnv->setSource1_Alpha(osg::TexEnvCombine::CONSTANT);
+        texEnv->setCombine_RGB(osg::TexEnvCombine::REPLACE);
+        texEnv->setSource0_RGB(osg::TexEnvCombine::PREVIOUS);
+        texEnv->setConstantColor(osg::Vec4f(1.0f, 1.0f, 1.0f, 1.0f));
```

Comment:
Thanks for finding that! That fixed it.

I think I'm just confused by how glColorMaterial works. I thought that using that would permit the diffuse color to be set via glColor for the entire material. I was under the impression that with GL_LIGHTING enabled, that it would ignore the vertex colors, and instead use the material colors with the texture colors in the lighting equation to shade it properly.

I guess I need to read that documentation again!